sycl/source/detail/memory_manager.cpp

```diff
@@ -9,17 +9,106 @@
 #include <CL/sycl/detail/memory_manager.hpp>
 #include <detail/context_impl.hpp>
 #include <detail/event_impl.hpp>
+#include <detail/mem_alloc_helper.hpp>
 #include <detail/queue_impl.hpp>
 
 #include <algorithm>
 #include <cassert>
 #include <cstring>
 #include <vector>
 
+#ifdef XPTI_ENABLE_INSTRUMENTATION
+#include <xpti/xpti_data_types.h>
+#include <xpti/xpti_trace_framework.hpp>
+#endif
+
 __SYCL_INLINE_NAMESPACE(cl) {
 namespace sycl {
 namespace detail {
 
+#ifdef XPTI_ENABLE_INSTRUMENTATION
+uint8_t GMemAllocStreamID;
+xpti::trace_event_data_t *GMemAllocEvent;
+#endif
+
+uint64_t emitMemAllocBeginTrace(uintptr_t ObjHandle, size_t AllocSize,
+                                size_t GuardZone) {
+  (void)ObjHandle;
+  (void)AllocSize;
+  (void)GuardZone;
+  uint64_t CorrelationID = 0;
+#ifdef XPTI_ENABLE_INSTRUMENTATION
+  if (xptiTraceEnabled()) {
+    xpti::mem_alloc_data_t MemAlloc{ObjHandle, 0 /* alloc ptr */, AllocSize,
+                                    GuardZone};
+
+    CorrelationID = xptiGetUniqueId();
+    xptiNotifySubscribers(
+        GMemAllocStreamID,
+        static_cast<uint16_t>(xpti::trace_point_type_t::mem_alloc_begin),
+        GMemAllocEvent, nullptr, CorrelationID, &MemAlloc);
+  }
+#endif
+  return CorrelationID;
+}
+
+void emitMemAllocEndTrace(uintptr_t ObjHandle, uintptr_t AllocPtr,
+                          size_t AllocSize, size_t GuardZone,
+                          uint64_t CorrelationID) {
+  (void)ObjHandle;
+  (void)AllocPtr;
+  (void)AllocSize;
+  (void)GuardZone;
+  (void)CorrelationID;
+#ifdef XPTI_ENABLE_INSTRUMENTATION
+  if (xptiTraceEnabled()) {
+    xpti::mem_alloc_data_t MemAlloc{ObjHandle, AllocPtr, AllocSize, GuardZone};
+
+    xptiNotifySubscribers(
+        GMemAllocStreamID,
+        static_cast<uint16_t>(xpti::trace_point_type_t::mem_alloc_end),
+        GMemAllocEvent, nullptr, CorrelationID, &MemAlloc);
+  }
+#endif
+}
+
+uint64_t emitMemReleaseBeginTrace(uintptr_t ObjHandle, uintptr_t AllocPtr) {
+  (void)ObjHandle;
+  (void)AllocPtr;
+#ifdef XPTI_ENABLE_INSTRUMENTATION
+  uint64_t CorrelationID = 0;
+  if (xptiTraceEnabled()) {
+    xpti::mem_alloc_data_t MemAlloc{ObjHandle, AllocPtr, 0 /* alloc size */,
+                                    0 /* guard zone */};
+
+    CorrelationID = xptiGetUniqueId();
+    xptiNotifySubscribers(
+        GMemAllocStreamID,
```

Comment:
It looks like **GMemAllocStreamID** is not initialized anywhere. is that expected?